phoenix-core/src/main/java/org/apache/phoenix/expression/function/UDFExpression.java

```diff
@@ -0,0 +1,217 @@
+    }
+
+    @Override
+    public PDataType getDataType() {
+        return udfFunction.getDataType();
+    }
+
+    @Override
+    public String getName() {
+        return udfFunction.getName();
+    }
+
+    @Override
+    public OrderPreserving preservesOrder() {
+        return udfFunction.preservesOrder();
+    }
+
+    @Override
+    public KeyPart newKeyPart(KeyPart childPart) {
+        return udfFunction.newKeyPart(childPart);
+    }
+
+    @Override
+    public int getKeyFormationTraversalIndex() {
+        return udfFunction.getKeyFormationTraversalIndex();
+    }
+
+    @Override
+    public void write(DataOutput output) throws IOException {
+        super.write(output);
+        WritableUtils.writeString(output, tenantId.getString());
+        WritableUtils.writeString(output, this.functionClassName);
+        if(this.jarPath == null) {
+            WritableUtils.writeString(output, "");
+        } else {
+            WritableUtils.writeString(output, this.jarPath);
+        }
+    }
+    
+    @Override
+    public void readFields(DataInput input) throws IOException {
+        super.readFields(input);
+        this.tenantId = PNameFactory.newName(WritableUtils.readString(input));
+        this.functionClassName = WritableUtils.readString(input);
+        String str = WritableUtils.readString(input);
+        this.jarPath = str.length() == 0 ? null: str;
+        constructUDFFunction();
+    }
+
+    private void constructUDFFunction() {
+        try {
+            DynamicClassLoader classLoader = getClassLoader(this.tenantId, this.jarPath);
+            Class<?> clazz = classLoader.loadClass(this.functionClassName);
+            Constructor<?> constructor = clazz.getConstructor(List.class);
+            udfFunction = (ScalarFunction)constructor.newInstance(this.children);
+        } catch (ClassNotFoundException | NoSuchMethodException | SecurityException
+                | InstantiationException | IllegalAccessException | IllegalArgumentException
+                | InvocationTargetException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static DynamicClassLoader getClassLoader(final PName tenantId, final String jarPath) {
+        DynamicClassLoader cl = tenantIdSpecificCls.get(tenantId);
+        String parent = null;
+        if (cl != null) return cl;
+        if(jarPath != null && !jarPath.isEmpty()) {
+            cl = pathSpecificCls.get(jarPath);
+            if (cl != null) return cl;
+            Path path = new Path(jarPath);
+            if(jarPath.endsWith(".jar")) {
+                parent = path.getParent().toString();
+            } else {
+                parent = path.toString();
+            }
+        }
+        if (jarPath == null || jarPath.isEmpty() || config.get(DYNAMIC_JARS_DIR_KEY) != null
+                && (parent != null && parent.equals(config.get(DYNAMIC_JARS_DIR_KEY)))) {
+            Lock lock = locker.acquireLock(tenantId.getString());
```

Comment:
Rajesh says: Just want to clarify one thing here the map is concurrent map so get and putifAbsent will be atomic but there is a chance like between get and put if other client already puts a class loader for the tenant then the classloader and configuration object created become garbage. That's why using locker. If that's ok I can do as you suggested. What do u say?

Samarth says: How expensive is creating a new classloader and config? To me it doesn't look too expensive. So I wouldn't worry about creating and discarding the two. Besides that race condition of another thread winning the race between get() and putIfAbsent is not very likely. So to sum it up, yes, I would get rid of the KeyLocker altogether.